src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,494 @@
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).thenExpand(info, 32);
+ *
+ *}
+ *
+ * @see javax.crypto.KDF
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public interface HKDFParameterSpec extends AlgorithmParameterSpec {
+
+    /**
+     * This {@code Builder} builds {@code Extract} and {@code ExtractThenExpand}
+     * objects.
+     * <p>
+     * The {@code Builder} is initialized via the {@code ofExtract} method of
+     * {@code HKDFParameterSpec}. As stated in the class description,
+     * {@code addIKM} and/or {@code addSalt} may be called as needed. Finally,
+     * the object is "built" by calling either {@code extractOnly} or
+     * {@code thenExpand} for {@code Extract} and {@code ExtractThenExpand}
+     * use-cases respectively. Note that the {@code Builder} is not
+     * thread-safe.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class Builder {
+
+        List<SecretKey> ikms = new ArrayList<>();
+        List<SecretKey> salts = new ArrayList<>();
+
+        private Builder() {}
+
+        /**
+         * Creates a {@code Builder} for an {@code Extract}.
+         *
+         * @return a {@code Builder} to mutate
+         */
+        private Builder createBuilder() {
+            return this;
+        }
+
+        /**
+         * Builds an {@code Extract} from the current state of the
+         * {@code Builder}.
+         *
+         * @return an immutable {@code Extract}
+         */
+        public Extract extractOnly() {
+            return new Extract(ikms, salts);
+        }
+
+        /**
+         * Builds an {@code ExtractThenExpand}.
+         *
+         * @param info
+         *     the optional context and application specific information (may be
+         *     {@code null}); the byte array is copied to prevent subsequent
+         *     modification
+         * @param length
+         *     the length of the output keying material (must be greater than 0)
+         *
+         * @implNote HKDF implementations will enforce that the length is less
+         * than 255 * HMAC length.
+         *
+         * @return an {@code ExtractThenExpand}
+         *
+         * @throws IllegalArgumentException
+         *     if {@code length} is not greater than 0
+         */
+        public ExtractThenExpand thenExpand(byte[] info, int length) {
+            return new ExtractThenExpand(
+                extractOnly(), info,
+                length);
+        }
+
+        /**
+         * Adds input keying material to the builder.
+         * <p>
+         * {@code addIKM} may be called when the input keying material value is to
+         * be assembled piece-meal or if part of the IKM is to be supplied by a
+         * hardware crypto device. This method appends to the existing list of
+         * values or creates a new list if there is none yet.
```

Comment:
Precisely, this should be "multiple addIKM may be called when..." since even in the simplest case you still need to call this method once. In fact, I don't think we need to talk about this detail every time. If you really want to emphasize this, create a sub-heading in the class spec, and inside each method just say "This method can be called multiple times on a single Builder object. See class_section_name".

Appending to or creating a list is implementation detail.

Reply:
I still think appending to or creating a list is implementation detail.

Reply:
The usage of "list" in specification could imply the `List` API, so it is best to avoid it as this could be interpreted as a requirement of the implementation. See for example the methods in the [`Collections` ](https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/util/Collections.html) API that operate on `List` objects but use "list" in the specification when referencing it.

If you want to be more abstract but still imply some sort of aggregation, I would suggest something like:

"The `ikms()` method of the `Extract` or `ExtractThenExpand` object that is subsequently built returns the assembled input keying material as a list of `SecretKey` objects."